lnwallet/rpcwallet/rpcwallet.go

```diff
@@ -401,48 +219,146 @@ func (r *RPCKeyRing) FinalizePsbt(packet *psbt.Packet, accountName string) error
+// TODO(guggero): Refactor btcwallet to accept ComputeInputScript as a function
+// parameter in FinalizePsbt so we can get rid of this code duplication.
+func (r *RPCKeyRing) FinalizePsbt(packet *psbt.Packet, _ string) error {
+	// Let's check that this is actually something we can and want to sign.
+	// We need at least one input and one output.
+	err := psbt.VerifyInputOutputLen(packet, true, true)
+	if err != nil {
+		return err
+	}
+
+	// Go through each input that doesn't have final witness data attached
+	// to it already and try to sign it. We do expect that we're the last
+	// ones to sign. If there is any input without witness data that we
+	// cannot sign because it's not our UTXO, this will be a hard failure.
+	tx := packet.UnsignedTx
+	sigHashes := txscript.NewTxSigHashes(tx)
+	for idx, txIn := range tx.TxIn {
+		in := packet.Inputs[idx]
+
+		// We can only sign if we have UTXO information available. We
+		// can just continue here as a later step will fail with a more
+		// precise error message.
+		if in.WitnessUtxo == nil && in.NonWitnessUtxo == nil {
+			continue
+		}
 
-	ctxt, cancel := context.WithTimeout(context.Background(), r.rpcTimeout)
-	defer cancel()
+		// Skip this input if it's got final witness data attached.
+		if len(in.FinalScriptWitness) > 0 {
+			continue
+		}
 
-	// We need to keep the local and remote wallet in sync. That's why we
-	// first attempt to also derive the next key on the remote wallet.
-	remoteDesc, err := r.walletClient.DeriveNextKey(ctxt, &walletrpc.KeyReq{
-		KeyFamily: int32(keyFam),
-	})
-	if err != nil {
-		return keychain.KeyDescriptor{}, fmt.Errorf("error deriving "+
-			"key on remote signer instance: %v", err)
+		// We can only sign this input if it's ours, so we try to map it
+		// to a coin we own. If we can't, then we'll continue as it
+		// isn't our input.
+		utxo, err := r.FetchInputInfo(&txIn.PreviousOutPoint)
+		if err != nil {
+			continue
+		}
+		fullTx := utxo.PrevTx
+		signDesc := &input.SignDescriptor{
+			KeyDesc: keychain.KeyDescriptor{},
+			Output: &wire.TxOut{
+				Value:    int64(utxo.Value),
+				PkScript: utxo.PkScript,
+			},
+			HashType:   in.SighashType,
+			SigHashes:  sigHashes,
+			InputIndex: idx,
+		}
+
+		// Find out what UTXO we are signing. Wallets _should_ always
+		// provide the full non-witness UTXO for segwit v0.
+		var signOutput *wire.TxOut
+		if in.NonWitnessUtxo != nil {
+			prevIndex := txIn.PreviousOutPoint.Index
+			signOutput = in.NonWitnessUtxo.TxOut[prevIndex]
+
+			if !psbt.TxOutsEqual(signDesc.Output, signOutput) {
+				return fmt.Errorf("found UTXO %#v but it "+
+					"doesn't match PSBT's input %v",
+					signDesc.Output, signOutput)
+			}
+
+			if fullTx.TxHash() != txIn.PreviousOutPoint.Hash {
+				return fmt.Errorf("found UTXO tx %v but it "+
+					"doesn't match PSBT's input %v",
+					fullTx.TxHash(),
+					txIn.PreviousOutPoint.Hash)
+			}
+		}
+
+		// Fall back to witness UTXO only for older wallets.
+		if in.WitnessUtxo != nil {
```

Comment:
I guess when `in.WitnessUtxo` and `in.NonWitnessUtxo` are both non-nil, we will always end up here and overwrite the `signOutput`?

Reply:
Yes, that's correct. `in.WitnessUtxo` is the "most correct" one so we take that if available and overwrite any other value we might have.